Add support for hubspot single email send via marketing/transactional endpoint.

sblackstone opened the pull request:
This PR adds support for a using  the single email send API.

https://developers.hubspot.com/docs/api/marketing/transactional-emails 

This maps to a `POST /marketing/v3/transactional/single-email/send`

Usage Example:

```go
	props := make(map[string]string)
	props["favoritecolor"] = "green"
	res2, err := client.Marketing.Transactional.SendSingleEmail(&hs.SendSingleEmailProperties{
		EmailId: "xxxxxxxxxxxx",
		Message: &hs.SendSingleEmailMessage{
			To:     "testing@example.com",
			SendId: id.String(),
		},
		ContactProperties: &hs.Contact{
			FirstName: hs.NewString("Stephen"),
		},
		CustomProperties: props,
	})
```

On a successful request, this will return:

```go
&{RequestedAt:2023-01-31T20:34:26.575Z StatusId:Hv-ltMeaTXqm_kmxDE0w8AAAAYYJiboWAWm-0Q== Status:PENDING}
```

ttyfky commented on marketing_transactional.go:
```diff
@@ -0,0 +1,52 @@
+
+const transactionalBasePath = "transactional"
+
+// DealService is an interface of deal endpoints of the HubSpot API.
```
Comment here and one for TransactionalServiceOp seem to be copied from DealService. Could you make these comments fit for Transactional Service?
As you may know, GoDoc starts from the target resource name (i.e. `TransactionalService` in this case) appreciate if you can follow this practice!

sblackstone replied:
Comments fixed.

ttyfky commented on marketing.go:
```diff
@@ -1,13 +1,23 @@
 type Marketing struct {
-	Email MarketingEmailService
+	Email         MarketingEmailService
+	Transactional TransactionalServiceOp
```
```suggestion
	Transactional TransactionalService
```

ttyfky replied:
Here suppose to have an interface the same as `MarketingEmailService`.

sblackstone replied:
This should be fixed.

sblackstone replied:
FWIW, I don't think theres a huge case for an interface here as no other object is ever going to reimplement these things.  Test cases could cover the concern that a function gets accidentally renamed or un-exposed.

OTOH, these functions are very difficult to write tests for without a dedicated HubSpot account for that purpose or mocking out their endpoints.

I'm happy to match style, so this is all set and its exposing the interface.

ttyfky replied:
Thank you for the update.  The intention to keep this field as interface is let this mockable. 
We can mock in http client layer to prevent external API call, but we also keep the capability to replace implementation by mock in this layer.

ttyfky commented on marketing_transactional.go:
```diff
@@ -0,0 +1,52 @@
+	Status      string `json:"status"`
+}
+
+func (s *TransactionalServiceOp) SingleEmailSend(props *SingleSendProperties) (*SingleEmailSendResponse, error) {
```
Could you make this method as `SendSingleEmail`?  We didn't have clear document unfortunately, but we usually give 'verb + object` name to a method.

Another idea is make `TransactionalService` to `TransactionalEmailService`, and change this function to `Send`.

ttyfky commented on marketing.go:
```diff
@@ -1,13 +1,23 @@
+	Transactional TransactionalServiceOp
 }
 
+// "/marketing/v3/transactional/single-email/send"
```
GoDoc to explain this function is supposed  to come here. How about this comment to move to the above of definition  `fmt.Sprintf("%s/%s/%s", marketingBasePath, c.apiVersion, transactionalBasePath)`?

sblackstone replied:
This was unintended, removed.

ttyfky commented:
Thank you for the PR! I gave some comments so please take a look.

sblackstone commented:
Addressed all requested changes with fixes.

Updated sample at the top of this PR to reflect naming schemes.

Re-ran tests against API and everything appears to still be working.

kk-no commented on marketing_transactional.go:
```diff
@@ -0,0 +1,50 @@
+	From    string   `json:"from,omitempty"`
+	SendId  string   `json:"sendId,omitempty"`
+	ReplyTo []string `json:"replyTo,omitempty"`
+	Cc      string   `json:"cc,omitempty"`
```
From following the api document, `Cc` and `Bcc` are array types.
https://developers.hubspot.com/docs/api/marketing/transactional-emails#endpoint?spec=POST-/marketing/v3/transactional/single-email/send
```suggestion
	Cc      []string   `json:"cc,omitempty"`
	Bcc     []string   `json:"bcc,omitempty"`
```

kk-no commented on marketing_transactional.go:
```diff
@@ -0,0 +1,50 @@
+	CustomProperties  interface{}             `json:"customProperties,omitempty"`
+}
+
+type SendSingleEmailResponse struct {
```
Wasn't the Response in the following format?
https://developers.hubspot.com/docs/api/marketing/transactional-emails#endpoint?spec=POST-/marketing/v3/transactional/single-email/send
```
{
  "statusId": "string",
  "sendResult": "SENT",
  "requestedAt": "2023-02-01T02:26:03.209Z",
  "startedAt": "2023-02-01T02:26:03.209Z",
  "completedAt": "2023-02-01T02:26:03.209Z",
  "status": "PENDING",
  "eventId": {
    "created": "2023-02-01T02:26:03.209Z",
    "id": "3fa85f64-5717-4562-b3fc-2c963f66afa6"
  }
}
```

sblackstone replied:
Perhaps I'm missing something here - but I orginally tried using `ResponseResource` instead of defining my own, and decoding seemed to produce: 

```
&{ID: Archived:false Associations:<nil> Properties:<nil> CreatedAt:nil UpdatedAt:nil ArchivedAt:nil}
```

kk-no replied:
hmm.. I understood. thank you.

kk-no commented on marketing_transactional.go:
```diff
@@ -0,0 +1,50 @@
+	SendSingleEmail(props *SendSingleEmailProperties) (*SendSingleEmailResponse, error)
+}
+
+// Provides the default implementation of TransactionService.
```
It is recommended that godoc begin with a struct name. (or function name, interface name etc..)
```suggestion
// TransactionalServiceOp provides the default implementation of TransactionService.
```

kk-no reviewed (commented):
Thank you for the PR!
I have made some comments and would appreciate your confirmation.

kk-no commented on marketing_transactional.go:
```diff
@@ -0,0 +1,50 @@
+		return nil, err
+	}
+	return resource, nil
+
```
nits: If there is no particular preference, I want to delete the line breaks.

sblackstone replied:
Fixed - It's just my personal style, I'm happy to match the codebase's style.

kk-no commented on marketing_transactional.go:
```diff
@@ -0,0 +1,50 @@
+}
+
+type SendSingleEmailProperties struct {
+	EmailId           string                  `json:"emailId"`
```
Should `EmailID` be a string?
The api documentation says `int32`.

sblackstone replied:
Fixed - it seems we're using int64 everywhere else in the codebase so used that.   If you prefer it to use int32, I'm fine with that too.

Where are you looking for type documentation incidentally?   The doc page I used to work against didn't have any types - it just says "a json object with these properties"  

However I dug into their official TypeScript docs and that seems to match what you're saying here.

Thanks.

kk-no replied:
I was referring to the document below. The details of each API are described in the "Endpoints" tab.
There is no problem with the type being `int64`. thank you.

https://developers.hubspot.com/docs/api/marketing/transactional-emails#endpoint?spec=POST-/marketing/v3/transactional/single-email/send

![image](https://user-images.githubusercontent.com/49366645/216536373-8b81a930-9ddb-4a2a-9c91-fc7f0d92e96a.png)

kk-no reviewed (commented):
Confirmed the fix. 
I made a new comment, so please check it.

kk-no reviewed (approved):
@sblackstone 
Thank you for your contribution!
Please wait as we will release a new version soon.
